src/codewhisperer/service/keyStrokeHandler.ts

```diff
@@ -42,6 +42,49 @@ export class KeyStrokeHandler {
         return (this.#instance ??= new this())
     }
 
+    public startIdleTimeTriggerTimer(
+        event: vscode.TextDocumentChangeEvent,
+        editor: vscode.TextEditor,
+        client: DefaultCodeWhispererClient,
+        config: ConfigurationEntry
+    ) {
+        if (this.idleTriggerTimer) {
+            clearInterval(this.idleTriggerTimer)
+            this.idleTriggerTimer = undefined
+        }
+        if (!this.shouldTriggerIdleTime()) {
+            return
+        }
+        this.idleTriggerTimer = setInterval(() => {
+            const duration = (performance.now() - RecommendationHandler.instance.lastInvocationTime) / 1000
+            if (duration < CodeWhispererConstants.invocationTimeIntervalThreshold) {
+                return
+            }
+
+            try {
+                this.invokeAutomatedTrigger('IdleTime', editor, client, config)
+            } finally {
+                if (this.idleTriggerTimer) {
+                    clearInterval(this.idleTriggerTimer)
+                    this.idleTriggerTimer = undefined
+                }
+            }
+        }, CodeWhispererConstants.idleTimerPollPeriod)
+    }
+
+    public shouldTriggerIdleTime(): boolean {
+        if (isCloud9() && RecommendationHandler.instance.isGenerateRecommendationInProgress) {
+            return false
+        }
+        if (isInlineCompletionEnabled() && InlineCompletionService.instance.isPaginationRunning()) {
+            return false
+        }
+        if (InlineCompletion.instance.getIsActive || InlineCompletion.instance.isPaginationRunning()) {
+            return false
+        }
```

Comment:
nit: another way to do this is to assign each condition to a more descriptive name and check them all at the end